enterprise/server/remote_execution/containers/podman/podman.go

```diff
@@ -236,6 +245,31 @@ func (c *podmanCommandContainer) IsImageCached(ctx context.Context) (bool, error
 }
 
 func (c *podmanCommandContainer) PullImage(ctx context.Context, creds container.PullCredentials) error {
+	psi, _ := pullOperations.LoadOrStore(c.image, &pullStatus{&sync.RWMutex{}, false})
+	ps, ok := psi.(*pullStatus)
+	if !ok {
+		alert.UnexpectedEvent("psi cannot be cast to *pullStatus")
+		return status.InternalError("PullImage failed: cannot get pull status")
+	}
+
+	ps.mu.RLock()
+	alreadyPulled := ps.pulled
+	ps.mu.RUnlock()
+
+	if alreadyPulled {
```

Comment:
We are using "podman pull" to check the credentials. If we are going to cache the pull results, then we need to handle creds auth separately (maybe using skokpeo inspect) ; or we have to add creds as part of the key. So when there are two pulls downloading the same images with different creds, we are not de-duped. 

In this code, we are trying to only have one podman pull process downloading the image and blocking other podman  commands that try to pull the same image. You actually pointed out an issue where if the cache is expired, then the following podman pulling actually might make network calls at the same time. But the alternative might be to remove the "alreadyPulled" boolean and always have one podman pull process downloading the same image (code: https://go.dev/play/p/dVmxvhgB2Jy). This might block unnecessarily; and make the common cases slower.